data_capture/spreadsheet_converter.py

```
@@ -0,0 +1,132 @@
+
+        # skip the heading row, process the rest
+        for rx in range(1, sheet.nrows):
+            row = [None] * len(self.xl_heading_to_csv_idx_map)  # init row
+
+            for heading, xl_idx in heading_indices.items():
+                # item_val = cval(xl_idx)
+                cell = sheet.cell(rx, xl_idx)
+
+                cell_type = cell.ctype
+                cell_value = cell.value
+
+                csv_col_idx = self.xl_heading_to_csv_idx_map[heading]
+
+                if cell_type is XL_CELL_DATE:
+                    # convert to mm/dd/YYYY string
+                    date = xldate_as_datetime(cell_value, datemode)
+                    cell_value = date.strftime('%m/%d/%Y')
+
+                # Save the string value into the expected CSV col
+                # index of the row
+                row[csv_col_idx] = str(cell_value)
+
+            yield row
+
+        self.xls_file.seek(0)
+        return
+
+    def convert_file(self):
+        '''
+        Converts the input Region 10 XLS/X spreadsheet to a list
+        of CSV-like rows expected by contracts.loaders.region_10.Region10Loader
+        '''
+        return list(self.convert_next())
+
+    def get_heading_indices_map(self, raises=True):
+        '''
+        Given a sheet, returns a mapping of R10 Excel sheet headings
+        to the column indices associated with those fields in that sheet
+        '''
+        book = xlrd.open_workbook(file_contents=self.xls_file.read())
+        sheet = book.sheet_by_index(self.sheet_index)
+        headings = sheet.row(0)
+
+        idx_map = {}
+        for i, cell in enumerate(headings):
+            # find the val in the xl_heading_to_csv_idx_map
+            if cell.value in self.xl_heading_to_csv_idx_map:
+                idx_map[cell.value] = i
+
+        if raises and len(idx_map) != len(self.xl_heading_to_csv_idx_map):
+            raise ValueError('Missing expected column(s) in Excel sheet')
```

From a purely technical standpoint, I think that `ValueError` is fine, but I'm wondering if this is an error that we'd like to propagate back to the end-user if the file they've uploaded ends up throwing this error?  In that case we might want to subclass `ValueError`, catch it when calling this method from the view/form logic, and convert it into a Django `ValidationError` so that it makes its way to the end-user... On the other hand, if it's enough to just catch all exceptions and turn them into generic "Sorry buddy, your XLS file is messed up" errors then this is fine!